### ISSUE TITLE
Efficient SET_ROW and GET_ROWS operation

### ISSUE THREAD

**Issue by prajaramanTT** [tenstorrent/tt-metal#15942]
GGML uses GET_ROWS to get rows of tensors (used in KV Cache). Currently SET_ROW has
limited support via nlp_kv_cache_load_slice (tiled tensors, 3D only) but GET_ROWS has none
(note the S on the end, it has to support multiple rows at a time). This is also a major
performance pain as GGML copies the entire KV cache to CPU to let the CPU handle the
operation.
I think implementing this optimally is going to be a long discussion. IMO (correct me if I'm
wrong), TTNN natively uses tiled tensors. SET_ROW and GET_ROWS could be implemented with
row_major tensors. TTNN uses 64 element alignment for row major. Which is perfectly fine for
KV cache as it's common to have matrix width of at least 256. Plus due to GET_ROWS is
basically a gather operation with usually a very sparse index. Tiled tensors would simply
increase the memory bandwidth needed by unnecessary padding. Then the resulting row-major
tensor could be tilized separately.

**Comment by cmaryanTT** [comment]
@marty1885 - can you please give a precise definition of these operations. Thanks

**Comment by ntarafdar** [comment]
@yugi957 we can discuss this tomorrow, but I think this will be a wrapper for your pad-fill op modified to change a tensor within its shape (set-rows), and use @sjameelTT 's slice for get-rows

**Comment by marty1885** [comment]
@ntarafdar Any chance it works with row major tensors?

**Comment by zzigler-tt** [comment]
@ntarafdar can you please review and triage. Thanks

**Comment by marty1885** [comment]
For context currently set rows is implemented as the following TOSA hack. Get does not work remotely.

```c++
static bool ggml_backend_metalium_can_set_rows(const struct ggml_tensor * dst)
{
    // GGML has a weird order
    // result->src[0] = src
    // result->src[1] = idx
    // result->src[2] = dst
    const ggml_tensor *idxs = dst->src[1];
    // effectivly no-op
    if(idxs->ne[0] == 1 && idxs->ne[1] == 1 && idxs->ne[2] == 1 && idxs->ne[3] == 1 && ggml_n_dims(dst->src[0]) == 1) {
        return true;
    }

    const ggml_tensor* src = dst->src[0];
    if(is_integer_type(src->type)) {
        return false;
    }

    if(idxs->ne[2] == 1 && src->ne[3] == 1 && !is_view(idxs)) {
        return true;
    }

    return false;
}

static void ggml_backend_metalium_set_rows(ggml_backend_metalium_context * ctx, struct ggml_tensor * dst)
{
    GGML_UNUSED(ctx);
    GGML_METALIUM_OP_SANITY_CHECK(dst);
    GGML_METALIUM_OP_SRC0_SANITY_CHECK(dst);
    GGML_METALIUM_OP_SRC_SANITY_CHECK(dst, 2);

    ggml_tensor_extra_metalium* dst_meta = (ggml_tensor_extra_metalium*)dst->extra;
    ggml_tensor_extra_metalium* real_dst_meta = (ggml_tensor_extra_metalium*)dst->src[2]->extra;
    ggml_tensor_extra_metalium* idx_meta = (ggml_tensor_extra_metalium*)dst->src[1]->extra;

    auto real_dst = realize_ggml_view(dst->src[2]);
    auto src = realize_ggml_view(dst->src[0]);
    auto idx = idx_meta->tensor;
    const ggml_tensor *idxs = dst->src[1];

    // Setting on a 1 row tensor is guarenteed to just be a replacment
    if(idxs->ne[0] == 1 && idxs->ne[1] == 1 && idxs->ne[2] == 1 && idxs->ne[3] == 1 && ggml_n_dims(dst->src[2]) == 1) {
        *dst_meta = {
            .tensor = src,
        };
        *real_dst_meta = {
            .tensor = src,
        };
    }
    else {
        ggml_tensor_extra_metalium* idx_meta = (ggml_tensor_extra_metalium*)idxs->extra;
        GGML_ASSERT(idx_meta != nullptr);
        // The operation wants 3D tensor but we have 4D, op also wants index be 2d
        auto src3d = src->reshape(src->logical_shape().to_rank(3));
        auto idx2d = idx_meta->tensor->reshape(idx_meta->tensor->logical_shape().to_rank(2));
        auto real_dst3d = real_dst->reshape(real_dst->logical_shape().to_rank(3));
        ttnn::Tensor res = ttnn::tosa_scatter(real_dst3d, ttnn::tilize_with_zero_padding(idx2d), src3d, std::nullopt);
        fmt::println("res: {}", res.logical_shape());
        res = res.reshape(res.logical_shape().to_rank(4));
        *dst_meta = {
            .tensor = std::make_shared<ttnn::Tensor>(res),
        };
        *real_dst_meta = {
            .tensor = std::make_shared<ttnn::Tensor>(res),
        };
    }
}

```
